### DISCUSSION TITLE
One Next.js application Docker image for multiple environments

### DISCUSSION THREAD

**Discussion by flaksp** [vercel/next.js#27145] [Help]
Hello everyone!

We run our apps in Kubernetes, so we are required to use Docker to serve our Next.js app. We store pre-release code in the `main` Git branch, and when we are ready to publish our app, we create a Git tag that triggers the Docker image build.

The problem is we have a bunch of API test/dev/stage environments we want to test our Next.js app against. So it would be cool for us to use a single Docker image with our Next.js app inside and be able to run it with different values for `API_BASE_HOST` environment variables (however, we have many more variables that are consumed by the browser, Node.js, or both).

The problems we were faced with:

* If we run `next build` at the Docker image build stage (via `RUN next build` inside Dockerfile), it will generate static pages only for a specific API environment.

  Environment variables for browser defined via `publicRuntimeConfig` Next.js config option or `NEXT_PUBLIC_` will be hardcoded during `next build` and we won't be able to change them when deploying an image into another environment that requires different values for such variables.

  Such image won't be reusable.
* We tried moving `next build` into `ENTRYPOINT`, so Next.js gets built when the container starts, but it's very slow and consumes a lot of memory, therefore we dropped this idea.

  However, we can reuse such images.
  
Could you give me some advice for such case? What does your flow look like when you work with multiple environments?

**Comment by sherifabdlnaby** [comment]
I've been looking for any officially supported method for doing so with no luck.
I don't get community members who just reply with "Why are you not using Vercel ?", It's not really an option in my case. And NextJS natively lacks lots of features that would make it play nice with Kubernetes, Horizontal Scaling, and as in your question.. the Build-Once-Run-Many capability.

**Comment by sherifabdlnaby** [comment]
Other than the very slow startup time, it doesn't make sense to run a container with 10~15X The resources it needs to serve the traffic, just because it needs them to build.

I don't think this can be solved 100% from within the framework. NextJS is not React. You can't just go use `find` and `sed` and replace the subsutuited ENV Vars and you get a working build. No, it won't work. Next fetches data and evaluate expressions at build time.

Recently NextJS Introduced some enhancments to subsequent builds as it uses old build as a cache. This won't help either, if you have a parameter for an API_BACKEND, and this backend is used in getStaticPaths, you will still have to rebuild your whole site.

The only solution I came up with and is currently implementing is to run a pre-release pod that does the build and save it to S3. and modify the containers in Kubernetes to sync the build on startup time.
